system-contracts/contracts/EvmGasManager.sol

```diff
@@ -0,0 +1,142 @@
+pragma solidity ^0.8.0;
+
+import {ACCOUNT_CODE_STORAGE_SYSTEM_CONTRACT} from "./Constants.sol";
+
+// We consider all the contracts (including system ones) as warm.
+uint160 constant PRECOMPILES_END = 0xffff;
+
+// Denotes that passGas has been consumed
+uint256 constant INF_PASS_GAS = 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff;
+
+contract EvmGasManager {
+    // We need trust to use `storage` pointers
+    struct WarmAccountInfo {
+        bool isWarm;
+    }
+
+    struct SlotInfo {
+        bool warm;
+        uint256 originalValue;
+    }
+
+    // We dont care about the size, since none of it will be stored/pub;ushed anywya
+    struct EVMStackFrameInfo {
+        bool isStatic;
+        uint256 passGas;
+    }
+
+    // The following storage variables are not used anywhere explicitly and are just used to obtain the storage pointers
+    // to use the transient storage with.
+    mapping(address => WarmAccountInfo) private warmAccounts;
+    mapping(address => mapping(uint256 => SlotInfo)) private warmSlots;
+    EVMStackFrameInfo[] private evmStackFrames;
+
+    function tstoreWarmAccount(address account, bool isWarm) internal {
+        WarmAccountInfo storage ptr = warmAccounts[account];
+
+        assembly {
+            tstore(ptr.slot, isWarm)
+        }
+    }
+
+    function tloadWarmAccount(address account) internal returns (bool isWarm) {
+        WarmAccountInfo storage ptr = warmAccounts[account];
+
+        assembly {
+            isWarm := tload(ptr.slot)
+        }
+    }
+
+    function tstoreWarmSlot(address _account, uint256 _key, SlotInfo memory info) internal {
+        SlotInfo storage ptr = warmSlots[_account][_key];
+
+        bool warm = info.warm;
+        uint256 originalValue = info.originalValue;
+
+        assembly {
+            tstore(ptr.slot, warm)
+            tstore(add(ptr.slot, 1), originalValue)
+        }
+    }
+
+    function tloadWarmSlot(address _account, uint256 _key) internal view returns (SlotInfo memory info) {
+        SlotInfo storage ptr = warmSlots[_account][_key];
+
+        bool isWarm;
+        uint256 originalValue;
+
+        assembly {
+            isWarm := tload(ptr.slot)
+            originalValue := tload(add(ptr.slot, 1))
+        }
+
+        info.warm = isWarm;
+        info.originalValue = originalValue;
+    }
+
+    modifier onlySystemEvm() {
+        require(ACCOUNT_CODE_STORAGE_SYSTEM_CONTRACT.isAccountEVM(msg.sender), "only system evm");
+        _;
+    }
+
+    /*
+        returns true if the account was already warm
+    */
+    function warmAccount(address account) external payable onlySystemEvm returns (bool wasWarm) {
+        if (uint160(account) < PRECOMPILES_END) return true;
+
+        wasWarm = tloadWarmAccount(account);
+        if (!wasWarm) tstoreWarmAccount(account, true);
+    }
+
+    function isSlotWarm(uint256 _slot) external view returns (bool) {
+        return tloadWarmSlot(msg.sender, _slot).warm;
```

Comment:
I would prefer to add a single-`tload` implementation here to reduce the number of queries to transient storage. This can reduce the load on the prover